lib/Sema/CSGen.cpp

```diff
@@ -3667,35 +3667,54 @@ void ConstraintSystem::optimizeConstraints(Expr *e) {
   e->walk(optimizer);
 }
 
-bool swift::isExtensionApplied(DeclContext &DC, Type BaseTy,
-                               const ExtensionDecl *ED) {
-  if (!ED->isConstrainedExtension() ||
-      // We'll crash if we leak type variables from one constraint
+bool swift::isGenericContextApplied(DeclContext *DC, Type BaseTy,
+                                    GenericContext *GC) {
+  if (// We'll crash if we leak type variables from one constraint
       // system into the new one created below.
       BaseTy->hasTypeVariable() ||
       // We can't do anything if the base type has unbound generic
       // parameters either.
       BaseTy->hasUnboundGenericType())
     return true;
 
-  TypeChecker *TC = &createTypeChecker(DC.getASTContext());
-  TC->validateExtension(const_cast<ExtensionDecl *>(ED));
+  TypeChecker *TC = &createTypeChecker(DC->getASTContext());
+  DeclContext *baseDC = nullptr;
+  bool isExtension = false;
+
+  if (auto *ED = dyn_cast_or_null<ExtensionDecl>(GC->getAsDecl())) {
+    if (!ED->isConstrainedExtension())
+      return true;
+    TC->validateExtension(ED);
+    baseDC = ED->getExtendedNominal();
+    isExtension = true;
+  } else if (auto *VD = dyn_cast_or_null<ValueDecl>(GC->getAsDecl())) {
+    TC->validateDeclForNameLookup(VD);
+    if (!GC->getGenericSignature())
+      return true;
+    baseDC = VD->getDeclContext();
+  } else {
+    // Unknown generic context? Assume it applicable.
+    return true;
+  }
 
   ConstraintSystemOptions Options;
-  ConstraintSystem CS(*TC, &DC, Options);
+  ConstraintSystem CS(*TC, DC, Options);
   auto Loc = CS.getConstraintLocator(nullptr);
 
   // Prepare type substitution map.
-  SubstitutionMap Substitutions = BaseTy->getContextSubstitutionMap(
-    DC.getParentModule(), ED);
+  SubstitutionMap Substitutions =
+      BaseTy->getMetatypeInstanceType()->getContextSubstitutionMap(
+          DC->getParentModule(), baseDC);
 
   // For every requirement, add a constraint.
-  for (auto Req : ED->getGenericRequirements()) {
+  for (auto Req : GC->getGenericRequirements()) {
     if (auto resolved = Req.subst(Substitutions)) {
       CS.addConstraint(*resolved, Loc);
-    } else {
+    } else if (isExtension) {
       return false;
     }
+    // Unresolved requirements a requirement of the function itself. this does
+    // not prevent it from being applied. E.g. func foo<T: Sequence>(x: T).
```

Comment:
If you don't want subst() to fail, use getMemberSubstitutionMap() instead of getContextSubstitutionMap().

Reply:
If we use `getMemberSubstitutionMap()` for member with "same type" requirement like
`func foo<U>(x: U) where T == Int {}`, the map returned from `getMemberSubstitutionMap()`  substitute `T` with `Int`. So the requirement turns to `Int == Int` which always success.